File: 2958. Length of Longest Subarray With at Most K Frequency/Solution.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        if k < 1 : return 0
        ans = right = left = 0
        freq = defaultdict(int)

        while right < len(nums):
            key = nums[right]
            freq[key] += 1

            while left <= right and freq[key] > k:
                freq[nums[left]] -= 1
                left += 1

            ans = max(ans, right - left + 1)
            right += 1

        return ans

File: 2958. Length of Longest Subarray With at Most K Frequency/Solution.java
class Solution {
    public int maxSubarrayLength(int[] nums, int k) {
        if(k<1) return 0;

        int n = nums.length;
        int left = 0;
        Map<Integer, Integer> freq = new HashMap<>();

        int max = 0;

        for(int right = 0; right < n; right++){

            int key = nums[right];

            freq.put(key, freq.getOrDefault(key,0)+1);

            while(left <= right && freq.get(key) > k){
                int leftKey = nums[left];
                freq.put(leftKey, freq.getOrDefault(leftKey,1)-1);
                left++;
            }

            max = Math.max(max, right - left +1);

        }

        return max;
    }
}